### ISSUE TITLE
Tactics in UniMath

### ISSUE THREAD

**Issue by siddharthist** [UniMath/UniMath#853]
As discussed on #828, we need to come up with a general policy for what kinds of tactics may be included and used in UniMath, and in what situations. 

I'll show my hand early: I generally support tactics, I think if UniMath is to become more widely used then automation is essential. The people who currently work with proof assistants are already used to the extra effort of doing trivial proofs, but mathematicians in general dislike this. 

Moreover, tactics that solve trivial goals seem useless to those who understand how to prove these goals already (see e.g. my proposed tactic `hleveltac` in #828), but more casual contributors might not know the names of these lemmas or where to find them. It lowers the barrier to entry to have them "packaged" in this way. 

I agree with @mortberg that we should be very careful about complex tactics that might generate non-optimal proof terms, and should not use tactics for goals where a specific _choice_ has to be made (such as proofs of contractibility), and that a reasonable compromise is only using heavy-handed tactics in opaque proofs (although the use of `Opaque` and `Qed` is also not agreed upon yet).

**Comment by mortberg** [comment]
I generally prefer type classes or canonical structures to obtain the kind of automation that a tactic like `hleveltac` provides. I have always felt like complicated special purpose tactics is the least elegant way to obtain automation in Coq. I think the two methods I mention are more fundamental and less ad hoc. In some sense they provide ways to let the user "teach" Coq basic mathematical facts and Coq can then use this knowledge to solve subgoals. I would definitely prefer if we added some simple type classes instead of adding more complicated tactics, however I'm not convinced that we need more automation at all...

**Comment by DanGrayson** [comment]
It was sound of Vladimir to insist we not use type classes or canonical structures, in case we move to another proof assistant in the future.

**Comment by mortberg** [comment]
Ah right, with that reasoning we should not rely on Ltac either. I think it is much more likely that future proof assistants have type classes. At the moment all of Agda, Coq Lean and Isabelle have type classes, but only Coq has Ltac...

**Comment by mortberg** [comment]
PS: It seems to me like the future for Ltac is a lot more uncertain than the future for type classes. I know there are at least two prototypes for replacements for Ltac out there (called Mtac and Mtac2), but I know of no proposed replacement for type classes...

**Comment by DanGrayson** [comment]
Ltac is a low-level construct that would be easy to implement in a new proof assistant.  Alternatively, we could replace every Ltac call by the term the tactic finds.

**Comment by mortberg** [comment]
No, Ltac could only be implemented in a proof assistant with a tactic language like Coq. In Agda for example there is no tactic language, hence no Ltac. However the kind of proof/term search that type classes provide can be implemented for any proof assistant (which is illustrated by the fact that the major proof assistants all have type classes).

**Comment by DanGrayson** [comment]
Well, sure, but I include the tactic language as part of Ltac and my statement applies to it.

I don't know anything about type classes, so you may have a point about that.

**Comment by mortberg** [comment]
To sum up my opinions about automation and UniMath: I prefer type classes to Ltac, but as Vladimir was against using type classes I don't think we should start using them yet. I always found it interesting that we can do so much in UniMath using very little automation, so I'm leaning against not introducing any more automation for the time being. If we ever feel that the lack of automation is stopping us from formalizing interesting mathematics we could considering using more, but at this time I don't really feel like I would benefit a lot from more automation (this might change if I start doing more algebra where reasoning about tons of algebraic equations by hand gets extremely tedious).

**Comment by benediktahrens** [comment]
On 12/19/2017 10:10 PM, Anders Mörtberg wrote:
> To sum up my opinions about automation and UniMath: I prefer type
> classes to Ltac, but as Vladimir was against using type classes I
> don't think we should start using them yet. I always found it
> interesting that we can do so much in UniMath using very little
> automation, so I'm leaning against not introducing any more
> automation for the time being. 

Be careful with the number of negations here :-)

> If we ever feel that the lack of
> automation is stopping us from formalizing interesting mathematics we
> could considering using more, but at this time I don't really feel
> like I would benefit a lot from more automation (this might change if
> I start doing more algebra where reasoning about tons of algebraic
> equations by hand gets extremely tedious).

Note that there are tactics for that in UniMath, that nobody has ever 
used - see package `Tactics`.

**Comment by DanGrayson** [comment]
Elementary reasoning about elementary things can get tedious quickly.  I found that out when I tried using |x-y| to make nat into a metric space.  Here's my tangled disgusting proof of the triangle inequality:
```
Definition nat_dist_trans x y z : nat_dist x z ≤ nat_dist x y + nat_dist y z.
Proof. intros. induction (natleorle x y) as [r|s].
       { rewrite (nat_dist_le _ _ r).
         induction (natleorle y z) as [t|u].
         { assert (u := istransnatgeh _ _ _ t r). rewrite (nat_dist_le _ _ t).
           rewrite (nat_dist_le _ _ u). apply (natlehandplusrinv _ _ x).
           rewrite (minusplusnmm _ _ u). rewrite (natpluscomm _ x).
           rewrite <- natplusassoc. rewrite (natpluscomm x).
           rewrite (minusplusnmm _ _ r). rewrite (natpluscomm y).
           rewrite (minusplusnmm _ _ t). apply isreflnatleh. }
         { rewrite (nat_dist_ge _ _ u).
           induction (natleorle x z) as [p|q].
           { rewrite (nat_dist_le _ _ p). apply (natlehandplusrinv _ _ x).
             rewrite (minusplusnmm _ _ p). rewrite natpluscomm.
             rewrite <- natplusassoc. rewrite (natpluscomm x).
             rewrite (minusplusnmm _ _ r). apply (natlehandplusrinv _ _ z).
             rewrite natplusassoc. rewrite (minusplusnmm _ _ u).
             apply (istransnatleh (m := y+z)).
             { apply natlehandplusr. exact u. }
             { apply natlehandplusl. exact u. } }
           { rewrite (nat_dist_ge _ _ q). apply (natlehandplusrinv _ _ z).
             rewrite (minusplusnmm _ _ q). rewrite natplusassoc.
             rewrite (minusplusnmm _ _ u). rewrite natpluscomm.
             apply (natlehandplusrinv _ _ x). rewrite natplusassoc.
             rewrite (minusplusnmm _ _ r). apply (istransnatleh (m := x+y)).
             { apply natlehandplusl. assumption. }
             { apply natlehandplusr. assumption. } } } }
       { rewrite (nat_dist_ge _ _ s).
         induction (natleorle z y) as [u|t].
         { assert (w := istransnatleh u s). rewrite (nat_dist_ge _ _ w).
           rewrite (nat_dist_ge _ _ u). apply (natlehandplusrinv _ _ z).
           rewrite (minusplusnmm _ _ w). rewrite natplusassoc.
           rewrite (minusplusnmm _ _ u). rewrite (minusplusnmm _ _ s).
           apply isreflnatleh. }
         { rewrite (nat_dist_le _ _ t).
           induction (natleorle x z) as [p|q].
           { rewrite (nat_dist_le _ _ p). apply (natlehandplusrinv _ _ x).
             rewrite (minusplusnmm _ _ p). apply (natlehandpluslinv _ _ y).
             rewrite (natplusassoc (x-y)). rewrite <- (natplusassoc y).
             rewrite (natpluscomm y (x-y)). rewrite (minusplusnmm _ _ s).
             apply (natlehandplusrinv _ _ y). rewrite (natplusassoc x).
             rewrite (natplusassoc _ x y). rewrite (natpluscomm x y).
             rewrite <- (natplusassoc _ y x). rewrite (minusplusnmm _ _ t).
             rewrite (natpluscomm z x). rewrite <- (natplusassoc x).
             rewrite (natplusassoc y). rewrite (natpluscomm z y).
             rewrite <- (natplusassoc y). apply (natlehandplusr _ _ z).
             apply (istransnatleh (m := x+y)).
             { apply natlehandplusr. assumption. }
             { apply natlehandplusl. assumption. } }
           { rewrite (nat_dist_ge _ _ q). apply (natlehandplusrinv _ _ z).
             rewrite (minusplusnmm _ _ q). apply (natlehandpluslinv _ _ y).
             rewrite (natplusassoc (x-y)). rewrite <- (natplusassoc y).
             rewrite (natpluscomm y (x-y)). rewrite (minusplusnmm _ _ s).
             apply (natlehandplusrinv _ _ y). rewrite (natplusassoc x).
             rewrite (natplusassoc _ z y). rewrite (natpluscomm z y).
             rewrite <- (natplusassoc _ y z). rewrite (minusplusnmm _ _ t).
             rewrite (natpluscomm y x). rewrite (natplusassoc x).
             apply natlehandplusl. apply (istransnatleh (m := z+y)).
             { apply natlehandplusr. assumption. }
             { apply natlehandplusl. assumption. } } } } Defined.
```

**Comment by siddharthist** [comment]
I think a lack of automation will never fully stop one from formalizing any part of mathematics, provided one has the time and energy to do so. For me, its a question of whether this library should be developed solely by those with the commitment and patience to learn its ins and outs and continue to prove trivial goals by hand, or focus on lowering barriers to contribution.

**Comment by siddharthist** [comment]
I agree that we could replace tactic-based proofs with the terms they generate upon moving to a new proof assistant. In fact, this will have to happen regardless of _which_ tactics we use, it is a consequence of using any tactics at all (even `refine` and `exact`). For this reason, I don't think portability is a strong argument against automation while Coq retains the `Print` command.

Also, we might ask that those that submit tactics commit to maintaining them in the face of changes to Ltac or a migration to Mtac. I would be very happy to learn Mtac and rewrite this tactic.

**Comment by DanGrayson** [comment]
Re: "For this reason, I don't think portability is a strong argument against automation while Coq retains the Print command."

Well, no: we don't want to replace a short human-understandable proof by a long confusing opaque one.

**Comment by siddharthist** [comment]
@DanGrayson With emphasis, 

> For this reason, I don't think _portability_ is a strong argument against automation while Coq retains the Print command.

I was trying to say that we shouldn't shun tactics _only_ because of the idea of moving to another proof assistant. There may be plenty of other arguments against automation, such as the one you mention, which I agree with.
